renku_notebooks/util/file_size.py

```diff
@@ -0,0 +1,58 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 - Swiss Data Science Center (SDSC)
+# A partnership between École Polytechnique Fédérale de Lausanne (EPFL) and
+# Eidgenössische Technische Hochschule Zürich (ETHZ).
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Helper utils for handling file size strings."""
+
+import re
+
+units = {
+    "b": 1,
+    "kb": 1000,
+    "mb": 1000 ** 2,
+    "gb": 1000 ** 3,
+    "tb": 1000 ** 4,
+    "m": 1000 ** 2,
+    "g": 1000 ** 3,
+    "t": 1000 ** 4,
+    "p": 1000 ** 5,
+    "e": 1000 ** 6,
+    "z": 1000 ** 7,
+    "y": 1000 ** 8,
+    "ki": 1024,
+    "mi": 1024 ** 2,
+    "gi": 1024 ** 3,
+    "ti": 1024 ** 4,
+    "pi": 1024 ** 5,
+    "ei": 1024 ** 6,
+    "zi": 1024 ** 7,
+    "yi": 1024 ** 8,
+}
+
+
+def parse_file_size(size_str):
```

Comment:
I am ok to do this in a separate PR. But I think it would look much nicer if we make a custom marshmallow field for memory/disk and give it a `bytes` property which will do this calc. This way the logic for validating the field and for extracting a number is all in one place. Also we could even potentially "deserialize" the string value to an integer that is just bytes - this can be done by overriding the `_deserialize` method.

We can do this by replacing this:
```
serverOptionDiskValue = fields.String(
    validate=lambda x: re.match(r"^(?:[1-9][0-9]*|[0-9]\.[0-9]*)[EPTGMK][i]{0,1}$", x)
    is not None,
    required=False,
)
```

By something like this:
```
class InformationAmount(fields.String):
    def _validate(self, value):
        return re.match(r"^(?:[1-9][0-9]*|[0-9]\.[0-9]*)[EPTGMK][i]{0,1}$", value) is not None

    @property
    def bytes(self):
         # insert logic from parse_file_size
```

Reply:
That's an excellent point - I was actually thinking how I should validate the string 🤦

Reply:
I'm not 100% sure where to go with this tbh, so I'll make an issue for it.

Reply:
Follow up: https://github.com/SwissDataScienceCenter/renku-notebooks/issues/582